crates/sargon/src/hierarchical_deterministic/bip39/bip39_seed.rs

```diff
@@ -36,59 +39,97 @@ impl BIP39Seed {
+                .map(|c| c.map_to_global_key_space()),
+        )
+        .secret_key()
+    }
+
+    pub fn derive_secp256k1_private_key(
+        &self,
+        hd_path: impl Into<HDPath>,
     ) -> Secp256k1PrivateKey {
-        let ck = self
-            .derive_slip10_private_key::<IotaSlip10Secp256k1::SecretKey, _>(
-                path.components.iter().cloned().map(|c| c.value),
-            );
-        Secp256k1PrivateKey::from_bytes(&*ck.secret_key().to_bytes())
+        let inner = self._derive_secp256k1_private_key(&hd_path.into());
+        Secp256k1PrivateKey::from_bytes(&*inner.to_bytes())
             .expect("Valid Secp256k1PrivateKey bytes")
         // `IotaSlip10Ed25519::SecretKey` implements `ZeroizeOnDrop` so should now be zeroized.
     }
 
+    pub fn derive_ed25519_private_key(
+        &self,
+        hd_path: impl Into<HDPath>,
+    ) -> Ed25519PrivateKey {
+        let inner = self._derive_ed25519_private_key(&hd_path.into());
+        Ed25519PrivateKey::from_bytes(inner.as_slice())
+            .expect("Valid Ed25519PrivateKey bytes")
+        // `IotaSlip10Ed25519::SecretKey` implements `ZeroizeOnDrop` so should now be zeroized.
+    }
+
+    pub fn derive_private_key_curve(
+        &self,
+        curve: SLIP10Curve,
+        path: impl Into<HDPath>,
+    ) -> PrivateKey {
+        match curve {
+            SLIP10Curve::Curve25519 => {
+                self.derive_ed25519_private_key(path).into()
+            }
+            SLIP10Curve::Secp256k1 => {
+                self.derive_secp256k1_private_key(path).into()
+            }
+        }
+    }
+
     pub fn derive_private_key<D>(
         &self,
         derivation: &D,
     ) -> HierarchicalDeterministicPrivateKey
     where
         D: Derivation,
     {
-        match derivation.curve() {
-            SLIP10Curve::Curve25519 => {
-                let key = self.derive_ed25519_private_key(derivation.hd_path());
-                HierarchicalDeterministicPrivateKey::new(
-                    key.into(),
-                    derivation.derivation_path(),
-                )
-            }
-            SLIP10Curve::Secp256k1 => {
-                let key =
-                    self.derive_secp256k1_private_key(derivation.hd_path());
-                HierarchicalDeterministicPrivateKey::new(
-                    key.into(),
-                    derivation.derivation_path(),
-                )
-            }
-        }
+        let key = self.derive_private_key_curve(
+            derivation.curve(),
+            derivation.derivation_path(),
+        );
+
+        HierarchicalDeterministicPrivateKey::new(
+            key,
+            derivation.derivation_path(),
+        )
     }
 }
 
+pub trait Derivation:
+    Clone + Into<DerivationPath> + HasDerivationPathSchemeObjectSafe
+{
+    fn derivation_path(&self) -> DerivationPath {
+        self.clone().into()
+    }
+}
+impl<T: Clone + Into<DerivationPath> + HasDerivationPathSchemeObjectSafe>
```

Comment:
This is a pattern we ought to start using, I recently learned it! Here we are saying that ALL types which impl `Clone + Into<DerivationPath> + HasDerivationPathSchemeObjectSafe` also impl `Derivation`